File: 剑指offer/剑指 Offer 24. 反转链表/reverseList.js
// 解法一:双指针
var reverseList = function (head) {
  let prev = null;
  let cur = head;
  while (cur !== null) {
    let temp = cur.next;
    cur.next = prev;
    prev = cur;
    cur = temp;
  }
  return prev;
};

//递归
var reverseList = function (head) {
  if (head === null || head.next === null) {
    return head;
  }
  const next = reverseList(head.next);
  head.next.next = head;
  head.next = null;
  return next;
};

File: 剑指offer/剑指 Offer 24. 反转链表/reverseList.py
# 双指针
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        prev, curr = None, head
        while curr != None:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp
        return prev


# 递归
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if head == None or head.next == None:
            return head
        next = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return next
